apps/omg_watcher/lib/omg_watcher/exit_processor_dispatcher.ex

```diff
@@ -0,0 +1,274 @@
+# Copyright 2019-2020 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.Watcher.ExitProcessorDispatcher do
+  @moduledoc """
+  Pretty sure it's doing some cool stuff.
+  """
+
+  require Logger
+
+  @doc """
+  Checks validity of all exit-related events and returns the list of actionable items.
+  Works with `OMG.State` to discern validity.
+
+  This function may also update some internal caches to make subsequent calls not redo the work,
+  but under unchanged conditions, it should have unchanged behavior from POV of an outside caller.
+  """
+  def check_validity() do
+    do_chack_validity()
+  end
+
+  def check_validity(timeout) do
+    do_chack_validity(timeout)
+  end
+
+  defp do_chack_validity(timeout \\ 5000) do
+    forwarded_result = forward(:check_validity, timeout)
+
+    Enum.reduce(forwarded_result, {:ok, []}, fn {chain_validity, events}, acc ->
+      {acc_chain_validity, acc_events} = acc
+
+      case acc_chain_validity do
+        :ok ->
+          {chain_validity, acc_events ++ events}
```

Comment:
a good practice is to avoid list concatenation this way. 
use [acc_events | events] and Enum.reverse when you're done.